lib/go-backoff/backoff.go

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package backoff
+
+/*
+ * Used to return an exponentially increasing time.Duration value that may be used
+ * to sleep between retries.
+ */
+
+import (
+	"github.com/apache/trafficcontrol/lib/go-log"
+	"math"
+	"math/rand"
+	"time"
+)
+
+// default minimum milliseconds is 100 msec
+const DefaultMinMs = time.Duration(100) * time.Millisecond
+
+// default maximum millseconds is 60000 (1m 0s)
+const DefaultMaxMs = time.Duration(60000) * time.Millisecond
+
+const DefaultFactor = float64(2)
+
+type BackOff struct {
+	attempt float64
+	Factor  float64
+	Min     time.Duration
+	Max     time.Duration
+}
+
+func init() {
+	rand.Seed(time.Now().UTC().UnixNano())
+}
+
+func NewBackOff(minMs time.Duration, maxMs time.Duration, factor float64) *BackOff {
+	min := minMs
+	max := maxMs
+	fac := factor
+
+	// verify arguments and set defaults if necessary.
+	if fac <= 1 {
+		log.Errorf("'The setting for 'Backoff.Factor' is too low, '%v', using default, '%v'\n", fac, DefaultFactor)
+		fac = DefaultFactor
+	}
+	if min < DefaultMinMs {
+		log.Errorf("'The setting for 'Backoff.Min' is too low, '%v', using default, '%v'\n", min, DefaultMinMs)
+		min = DefaultMinMs
```

Comment:
For a general utility package, it seems like it'd be better not to enforce specific minimum-minimums and maximum-maximums. Someone might want to use this for something very fast, or very slow.

If a specific app (like the Monitor below) should have a min-min or max-max, it's more flexible for this package to allow anything, and for the app itself to enforce the min/max.

It should probably validate min<=max though.